### PR TITLE
fix: stuck rollout when 2nd deployment happens before 1st finishes

### PR DESCRIPTION
**Pull request by ashutosh16**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

**Comment by ashutosh16** [comment]
@meeech  Would you be able to review the PR?

**Comment by meeech** [comment]
@ashutosh16 I'll have a look and test it out as well.

Q: Does this fix need to be done for the bluegreen as well. An identical change was made there.

**Comment by ashutosh16** [comment]
> @ashutosh16 I'll have a look and test it out as well.
> 
> Q: Does this fix need to be done for the bluegreen as well. An identical change was made there.

I tested with the following scenario:
- basic canary.  // worked with fix
- basic canary with analysis. // worked with fix
- basic canary with traffic router. //  worked as expected
- blue green //  worked as expected


the fix is not needed for b/g, since the svc is already update with correct `rollouts-pod-template-hash` when a new rs is created. Hence, isReplicaSetReferenced returns false, which means the svc is not referenced by any intermediary replication set

File: test/e2e/functional/rollout-canary-with-pause.yaml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            initialDelaySeconds: 10
+            httpGet:
+              path: /
+              port: 80
+            periodSeconds: 30
+  strategy:
+    canary:
+      steps:
+        - setWeight: 20
+        - pause: {}
```
Comment (meeech) [inline test/e2e/functional/rollout-canary-with-pause.yaml:75]: ~ have newline at the end of yaml files can help avoid pain.

File: test/e2e/canary_test.go [hunk 2]
```diff
@@ -565,6 +532,24 @@ func (s *CanarySuite) TestCanaryScaleDownOnAbortNoTrafficRouting() {
...
+		Then().
+		// Expect that the canary service selector has been moved back to stable
+		ExpectServiceSelector("rollout-canary-with-pause-canary", map[string]string{"app": "rollout-canary-with-pause"}, false).
+		When().
+		Sleep(3*time.Second).
+		Then().
+		ExpectRevisionPodCount("2", 0).
+		ExpectRevisionPodCount("3", 1))
+}
+
```
[the remark this answers is not in the document]
Comment (ashutosh16) [reply test/e2e/canary_test.go:552]: You're right the e2e does simulate the screnario correctly. I looked into  e2e function `UpdateSpec()` and it seems it resume the rollout  and then updating to new revision.   I'll fix it!

[the remark this answers is not in the document]
Comment (ashutosh16) [reply test/e2e/canary_test.go:584]: Fixed the testcase. It should fail without the fix.

**Review by zachaller — APPROVED** [review]

**Comment by meeech** [comment]
Linking back to the related issue: #3331

**Pull request merged**